File: app/src/main/java/pakutoma/miowidget/widget/SwitchWidget.java
package pakutoma.miowidget.widget;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.view.View;
import android.widget.RemoteViews;
import android.widget.Toast;

import java.util.Locale;

import pakutoma.miowidget.R;
import pakutoma.miowidget.service.GetTraffic;
import pakutoma.miowidget.service.SwitchCoupon;

/**
 * Implementation of App Widget functionality.
 */
public class SwitchWidget extends AppWidgetProvider {


    private static final String ACTION_GET_TRAFFIC = "pakutoma.miowidget.widget.SwitchWidget.ACTION_GET_TRAFFIC";
    private static final String ACTION_WIDGET_ENABLE = "pakutoma.miowidget.widget.SwitchWidget.ACTION_WIDGET_ENABLE";
    private static final String ACTION_WAIT_CHANGE_SWITCH = "pakutoma.miowidget.widget.SwitchWidget.ACTION_WAIT_CHANGE_SWITCH";

    private static final String ACTION_CALLBACK_GET_TRAFFIC = "pakutoma.miowidget.widget.SwitchWidget.ACTION_CALLBACK_GET_TRAFFIC";
    private static final String ACTION_CALLBACK_CHANGE_COUPON = "pakutoma.miowidget.widget.SwitchWidget.ACTION_CALLBACK_CHANGE_COUPON";

    private static Boolean isCouponEnable = true;

    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        if (intent.getAction().equals(ACTION_GET_TRAFFIC)) {
            setAlarm(context);
            Intent getTrafficIntent = new Intent(context, GetTraffic.class);
            context.startService(getTrafficIntent);
        }

        if (intent.getAction().equals(ACTION_CALLBACK_GET_TRAFFIC)) {
            updateTraffic(context,intent);
        }

        if (intent.getAction().equals(ACTION_CALLBACK_CHANGE_COUPON)) {
            changeSwitch(context,intent);
        }

        if (intent.getAction().equals(ACTION_WAIT_CHANGE_SWITCH)) {
            changeToWaitMode(context);
        }

        if (intent.getAction().equals(Intent.ACTION_BOOT_COMPLETED) || intent.getAction().equals(Intent.ACTION_MY_PACKAGE_REPLACED) || intent.getAction().equals(ACTION_WIDGET_ENABLE)) {
            setAlarm(context);
            Intent getTrafficIntent = new Intent(context, GetTraffic.class);
            context.startService(getTrafficIntent);
            Intent switchCouponIntent = new Intent(context, SwitchCoupon.class);
            context.startService(switchCouponIntent);
        }
    }

    private void setAlarm(Context context) {
        Intent alarmIntent = new Intent(context, SwitchWidget.class);
        alarmIntent.setAction(ACTION_GET_TRAFFIC);
        PendingIntent operation = PendingIntent.getBroadcast(context, 0, alarmIntent, 0);
        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        final long interval = 1000 * 60 * 5;
        final long nextAlarm = System.currentTimeMillis() + interval;
        am.set(AlarmManager.RTC, nextAlarm, operation);
    }

    private void cancelAlarm(Context context) {
        Intent alarmIntent = new Intent(context, SwitchWidget.class);
        alarmIntent.setAction(ACTION_GET_TRAFFIC);
        PendingIntent operation = PendingIntent.getBroadcast(context, 0, alarmIntent, 0);
        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        am.cancel(operation);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        setAlarm(context);
        Intent getTrafficIntent = new Intent(context, GetTraffic.class);
        context.startService(getTrafficIntent);
        Intent switchCouponIntent = new Intent(context, SwitchCoupon.class);
        context.startService(switchCouponIntent);
    }

    private void updateTraffic(Context context,Intent intent) {
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.switch_widget);
        if (!intent.getBooleanExtra("HAS_TOKEN",false)) {
            remoteViews.setTextViewText(R.id.data_traffic, "未認証");
        } else if (!intent.getBooleanExtra(("GET"),false)) {
            remoteViews.setTextViewText(R.id.data_traffic, "エラー");
        } else {
            int traffic = intent.getIntExtra("TRAFFIC",0);
            remoteViews.setTextViewText(R.id.data_traffic, convertPrefixString(traffic));
            isCouponEnable = intent.getBooleanExtra("COUPON",false);
            changeSwitchMode(remoteViews);
        }
        ComponentName thisWidget = new ComponentName(context, SwitchWidget.class);
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        manager.updateAppWidget(thisWidget, remoteViews);
    }

    private String convertPrefixString(int traffic) {
        String result;
        if (traffic < 1000) {
            result = String.format(Locale.US,"%dMB",traffic);
        } else if (traffic < 10000) {
            result = String.format(Locale.US,"%1$.2fGB",traffic / 1000.0);
        } else {
            result = String.format(Locale.US,"%1$.1fGB",traffic / 1000.0);
        }
        return result;
    }

    private void changeSwitch(Context context,Intent intent) {
        if (!intent.getBooleanExtra("HAS_TOKEN",false)) {
            Toast.makeText(context, "認証が行われていません。", Toast.LENGTH_SHORT).show();
            return;
        }
        if(intent.getBooleanExtra("CHANGE",false)) {
            isCouponEnable = intent.getBooleanExtra("NOW_SWITCH",false);
            Toast.makeText(context,"クーポンを" + (isCouponEnable ? "ON" : "OFF") + "に変更しました。",Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(context,"切り替えに失敗しました。",Toast.LENGTH_SHORT).show();
        }
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.switch_widget);
        changeSwitchMode(remoteViews);

        ComponentName thisWidget = new ComponentName(context, SwitchWidget.class);
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        manager.updateAppWidget(thisWidget, remoteViews);
    }

    private void changeToWaitMode(Context context) {
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.switch_widget);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_on_end, View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_start, View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_on_start, isCouponEnable ? View.VISIBLE : View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_end,isCouponEnable ? View.INVISIBLE : View.VISIBLE);
        ComponentName thisWidget = new ComponentName(context, SwitchWidget.class);
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        manager.updateAppWidget(thisWidget, remoteViews);
    }

    private void changeSwitchMode(RemoteViews remoteViews) {
        remoteViews.setViewVisibility(R.id.coupon_switch_base_on, isCouponEnable ? View.VISIBLE : View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_base, isCouponEnable ? View.INVISIBLE : View.VISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_on_end, isCouponEnable ? View.VISIBLE : View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_start, isCouponEnable ? View.INVISIBLE : View.VISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_on_start, View.INVISIBLE);
        remoteViews.setViewVisibility(R.id.coupon_switch_top_end, View.INVISIBLE);

    }

    @Override
    public void onEnabled(Context context) {

    }

    @Override
    public void onDisabled(Context context) {
        cancelAlarm(context);
        super.onDisabled(context);
    }
}

File: app/src/main/java/pakutoma/miowidget/widget/SwitchWidget.kt
package pakutoma.miowidget.widget

import android.annotation.TargetApi
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_CANCEL_CURRENT
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews

import java.util.Locale

import pakutoma.miowidget.R
import pakutoma.miowidget.service.FetchRemainsJobService
import pakutoma.miowidget.service.SwitchCouponService
import android.app.job.JobInfo
import android.app.job.JobScheduler
import android.app.NotificationManager
import android.app.NotificationChannel
import android.os.Build
import pakutoma.miowidget.service.FetchRemainsService
import java.util.concurrent.TimeUnit


/**
 * Implementation of App Widget functionality.
 */

class SwitchWidget : AppWidgetProvider() {
    companion object {
        private const val JOB_ID = 1
        private const val NOTIFICATION_CHANNEL_ID = "switch_service"
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        if (intent.action == Intent.ACTION_BOOT_COMPLETED) {
            startFetchRemainsService(context)
        }
        if(intent.action == Intent.ACTION_MY_PACKAGE_REPLACED) {
            registerFetchRemainsJobService(context)
            registerNotificationChannel(context)
            startFetchRemainsService(context)
        }
    }

    override fun onEnabled(context: Context) {
        super.onEnabled(context)
        registerFetchRemainsJobService(context)
        registerNotificationChannel(context)
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
        startFetchRemainsService(context)
    }

    override fun onDisabled(context: Context) {
        val jobScheduler = context.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
        jobScheduler.cancel(JOB_ID)
        super.onDisabled(context)
    }

    private fun registerNotificationChannel(context: Context) {
        @TargetApi(Build.VERSION_CODES.O)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            if(manager.getNotificationChannel(NOTIFICATION_CHANNEL_ID) == null) {
                val channel = NotificationChannel(
                        NOTIFICATION_CHANNEL_ID,
                        context.getString(R.string.switch_service),
                        NotificationManager.IMPORTANCE_MIN
                )
                manager.createNotificationChannel(channel)
            }
        }
    }

    private fun registerFetchRemainsJobService(context: Context) {
        val jobScheduler = context.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
        val jobs = jobScheduler.allPendingJobs
        if(jobs?.any { it.id == JOB_ID } != true) {
            val jobInfo = JobInfo.Builder(JOB_ID, ComponentName(context, FetchRemainsJobService::class.java))
                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
                    .setPeriodic(TimeUnit.MINUTES.toMillis(15))
                    .setPersisted(true)
                    .build()
            jobScheduler.schedule(jobInfo)
        }
    }

    private fun startFetchRemainsService(context: Context) {
        val intent = Intent(context, FetchRemainsService::class.java)
        @TargetApi(Build.VERSION_CODES.O)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent)
        } else {
            context.startService(intent)
        }
    }
}

fun changeToFetchingMode(context: Context) {
    val remoteViews = RemoteViews(context.packageName, R.layout.switch_widget)
    remoteViews.setTextViewText(R.id.data_traffic, "取得中")
    val thisWidget = ComponentName(context, SwitchWidget::class.java)
    val manager = AppWidgetManager.getInstance(context)
    manager.updateAppWidget(thisWidget, remoteViews)
}

fun changeToWaitMode(context: Context, isCouponEnable: Boolean) {
    val remoteViews = RemoteViews(context.packageName, R.layout.switch_widget)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_on_end, View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_start, View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_on_start, if (isCouponEnable) View.VISIBLE else View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_end, if (isCouponEnable) View.INVISIBLE else View.VISIBLE)
    val thisWidget = ComponentName(context, SwitchWidget::class.java)
    val manager = AppWidgetManager.getInstance(context)
    manager.updateAppWidget(thisWidget, remoteViews)
}

fun updateSwitchStatus(context: Context, hasToken: Boolean, couldGet: Boolean, remains: Int = -1, isCouponEnable: Boolean = false) {
    val remoteViews = RemoteViews(context.packageName, R.layout.switch_widget)
    remoteViews.setViewVisibility(R.id.coupon_switch_base_on, if (isCouponEnable) View.VISIBLE else View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_base, if (isCouponEnable) View.INVISIBLE else View.VISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_on_end, if (isCouponEnable) View.VISIBLE else View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_start, if (isCouponEnable) View.INVISIBLE else View.VISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_on_start, View.INVISIBLE)
    remoteViews.setViewVisibility(R.id.coupon_switch_top_end, View.INVISIBLE)
    if (!hasToken) {
        remoteViews.setTextViewText(R.id.data_traffic, "未認証")
    } else if (!couldGet) {
        remoteViews.setTextViewText(R.id.data_traffic, "エラー")
    } else {
        remoteViews.setTextViewText(R.id.data_traffic, convertPrefixString(remains))
    }
    val clickIntent = Intent(context, SwitchCouponService::class.java)
    val pendingIntent: PendingIntent
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        pendingIntent = PendingIntent.getForegroundService(context, 0, clickIntent, FLAG_CANCEL_CURRENT)
    } else {
        pendingIntent = PendingIntent.getService(context, 0, clickIntent, FLAG_CANCEL_CURRENT)
    }
    remoteViews.setOnClickPendingIntent(R.id.transparent_button, pendingIntent)
    val thisWidget = ComponentName(context, SwitchWidget::class.java)
    val manager = AppWidgetManager.getInstance(context)
    manager.updateAppWidget(thisWidget, remoteViews)
}

private fun convertPrefixString(traffic: Int): String {
    return when (traffic) {
        in 0..999 -> String.format(Locale.US, "%dMB", traffic)
        in 1000..10000 -> String.format(Locale.US, "%1$.2fGB", traffic / 1000.0)
        else -> String.format(Locale.US, "%1$.1fGB", traffic / 1000.0)
    }
}

fun setSwitchPendingIntent(context: Context) {
    val remoteViews = RemoteViews(context.packageName, R.layout.switch_widget)
    val clickIntent = Intent(context, SwitchCouponService::class.java)
    val pendingIntent: PendingIntent
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        pendingIntent = PendingIntent.getForegroundService(context, 0, clickIntent, FLAG_CANCEL_CURRENT)
    } else {
        pendingIntent = PendingIntent.getService(context, 0, clickIntent, FLAG_CANCEL_CURRENT)
    }
    remoteViews.setOnClickPendingIntent(R.id.transparent_button, pendingIntent)
    val thisWidget = ComponentName(context, SwitchWidget::class.java)
    val manager = AppWidgetManager.getInstance(context)
    manager.updateAppWidget(thisWidget, remoteViews)
}